File: Editorial/1001-1100/1055_Shortest_Way_to_Form_String/Go/2_Concatenate_until_Subsequence.go
func shortestWay(source string, target string) int {

    // Boolean array to mark all characters of source
    sourceChars := make([]bool, 26)
    for _, c := range source {
        sourceChars[c - 'a'] = true
    }

    // Check if all characters of target are present in source
    // If any character is not present, return -1
    for _, c := range target {
        if !sourceChars[c - 'a'] {
            return -1
        }
    }

    // Concatenate source until target is a subsequence of concatenated string
    concatenatedSource := source
    count := 1
    for !isSubsequence(target, concatenatedSource) {
        concatenatedSource += source
        count++
    }

    // Number of concatenations done
    return count
}

// To check if toCheck is a subsequence of inString
func isSubsequence(toCheck string, inString string) bool {
    i, j := 0, 0
    for i < len(toCheck) && j < len(inString) {
        if toCheck[i] == inString[j] {
            i++
        }
        j++
    }

    return i == len(toCheck)
}

File: Editorial/1001-1100/1055_Shortest_Way_to_Form_String/C/2_Concatenate_until_Subsequence.c
// To check if toCheck is a subsequence of inString
bool isSubsequence(char * toCheck, char * inString) {
    int i = 0, j = 0;
    while (toCheck[i] != '\0' && inString[j] != '\0') {
        if (toCheck[i] == inString[j]) {
            i++;
        }
        j++;
    }

    return toCheck[i] == '\0';
}

int shortestWay(char * source, char * target){
    
    // Boolean array to mark all characters of the source
    bool sourceChars[26] = {false};
    for (int i = 0; source[i] != '\0'; i++) {
        sourceChars[source[i] - 'a'] = true;
    }

    // Check if all characters of target are present in source
    // If any character is not present, return -1
    for (int i = 0; target[i] != '\0'; i++) {
        if (!sourceChars[target[i] - 'a']) {
            return -1;
        }
    }

    // Concatenate source until target is a subsequence of the concatenated string
    char concatenatedSource[1000001];
    strcpy(concatenatedSource, source);
    int count = 1;
    while (!isSubsequence(target, concatenatedSource)) {
        strcat(concatenatedSource, source);
        count++;
    }

    // Number of concatenations done
    return count;
}
